runtime/perfetto/db.go

```diff
@@ -164,14 +189,91 @@ func (d *DB) Store(ctx context.Context, app, version string, spans []sdktrace.Re
 }
 
 func (d *DB) storeEncoded(ctx context.Context, app, version string, encoded []byte) error {
-	const stmt = `
-		INSERT INTO traces(app, version, events)
-		VALUES (?,?,?);
-	`
-	_, err := d.execDB(ctx, stmt, app, version, string(encoded))
+	tx, err := d.db.BeginTx(ctx, &sql.TxOptions{Isolation: sql.LevelLinearizable})
 	if err != nil {
+		return err
+	}
+	defer tx.Rollback() //nolint:errcheck // rollback errors can be ignored
+
+	// Insert the encoded trace data.
+	const insertStmt = `
+		INSERT INTO traces(insertion_time, app, version, events)
+		VALUES (strftime('%Y-%m-%d %H:%M:%S:%f','now'),?,?,?);
+	`
+	if _, err := tx.ExecContext(ctx, insertStmt, app, version, string(encoded)); err != nil {
 		return fmt.Errorf("write trace to %s: %w", d.fname, err)
 	}
+
+	if d.opts.MaxTraceBytes > 0 {
+		// Update the size of trace data.
+		const updateSizeStmt = `UPDATE traces_size SET size = size + ? WHERE id = 0;`
+		if _, err := tx.ExecContext(ctx, updateSizeStmt, len(encoded)); err != nil {
+			return fmt.Errorf("update trace size in %s: %w", d.fname, err)
+		}
+	}
+
+	// See if we need to garbage-collect oldest traces.
+	if err := d.maybeGC(ctx, tx); err != nil {
+		return err
+	}
+
+	return tx.Commit()
+}
+
+func (d *DB) maybeGC(ctx context.Context, tx *sql.Tx) error {
+	if d.opts.MaxTraceBytes == 0 {
+		return nil
+	}
+
+	// Get the total size of trace data in DB.
+	const sizeQuery = `SELECT size FROM traces_size`
+	rows, err := tx.QueryContext(ctx, sizeQuery)
+	if err != nil {
+		return err
+	}
+	defer rows.Close()
+	var totalBytes int64
+	if rows.Next() {
+		if err := rows.Scan(&totalBytes); err != nil {
+			return err
+		}
+	}
+
+	if totalBytes < d.opts.MaxTraceBytes {
+		return nil
+	}
+
+	// Shrink trace data down to MaxTraceBytes / 2, to amortize the
+	// deletion cost.
+	toDeleteBytes := totalBytes - (d.opts.MaxTraceBytes / 2)
+
+	// Compute the trace length for every trace row, in insertion order.
+	const lenQuery = `SELECT rowid, length(events) FROM traces ORDER BY insertion_time`
```

Comment:
Could this be a massive temporary spike in resource usage (infrequent, but could hurt latency sensitive ops).

Reply:
Good point. Please look at the new solution which deletes a single row once the DB size crosses a threshold.